File: ClientAPI/src/main/java/pt/upskill/clientapi/Interfaces/CategoryDAO.java
package pt.upskill.clientapi.Interfaces;

import pt.upskill.clientapi.Models.Category;

import java.util.List;

/**
 * Data Access Object (DAO) interface for managing Category entities.
 * Provides methods to create, retrieve, and delete categories.
 */
public interface CategoryDAO {

    /**
     * Creates a new category in the system.
     *
     * @param category The category to be created.
     * @return The created category with any generated properties (e.g., ID).
     */
    Category createCategory(Category category);

    /**
     * Retrieves a category by its description.
     *
     * @param description The description of the category to retrieve.
     * @return The category with the specified description, or null if not found.
     */
    Category getCategoryByDescription(String description);

    /**
     * Retrieves a list of all categories in the system.
     *
     * @return A list of all categories.
     */
    List<Category> getAllCategories();

    /**
     * Deletes a category based on its description.
     *
     * @param description The description of the category to be deleted.
     */
    void deleteCategoryByDescription(String description);
}

File: ForumAPI/ForumAPI/Interfaces/CategoryDAO.cs
using ForumAPI.DTOs;
using ForumAPI.DTOs.CategoryDTOs;

namespace ForumAPI.Interfaces
{
    /// <summary>
    /// Interface that defines the data access operations for categories.
    /// </summary>
    public interface CategoryDAO
    {
        /// <summary>
        /// Retrieves a list of all categories.
        /// </summary>
        /// <returns>A task representing the asynchronous operation, with a list of <see cref="CategoryDTO"/> as the result.</returns>
        public Task<List<CategoryDTO>> GetAllCategories();

        /// <summary>
        /// Retrieves a category by its description.
        /// </summary>
        /// <param name="description">The description of the category to retrieve.</param>
        /// <returns>A task representing the asynchronous operation, with a <see cref="CategoryDTO"/> as the result.</returns>
        public Task<CategoryDTO> GetCategoryByDescription(string description);

        /// <summary>
        /// Deletes a category based on its description.
        /// </summary>
        /// <param name="description">The description of the category to delete.</param>
        /// <returns>A task representing the asynchronous operation.</returns>
        public Task DeleteCategory(string description);

        /// <summary>
        /// Creates a new category.
        /// </summary>
        /// <param name="categoryDTO">The category data transfer object containing the details of the category to be created.</param>
        /// <returns>A task representing the asynchronous operation, with a <see cref="ResponseMessage"/> as the result.</returns>
        public Task<ResponseMessage> CreateCategory(CategoryDTO categoryDTO);
    }
}